pkg/volumes/volumes.go

```diff
@@ -0,0 +1,126 @@
+// Copyright The Shipwright Contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package volumes
+
+import (
+	"fmt"
+
+	buildv1alpha1 "github.com/shipwright-io/build/pkg/apis/build/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+// StepVolumeMount describes volume mounts required by the build strategy
+// with corresponding fields necessary for the further checks
+type StepVolumeMount struct {
+	Name     string
+	ReadOnly bool
+}
+
+// NewStepVolumeMount creates an instance of the StepVolumeMount
+// structure
+func NewStepVolumeMount(name string, readOnly bool) StepVolumeMount {
+	return StepVolumeMount{
+		Name:     name,
+		ReadOnly: readOnly,
+	}
+}
+
+// TaskSpecVolumes creates a list of Volumes for the generated TaskSpec object and
+// checks for some erroneous situations around volumes and volume mounts
+func TaskSpecVolumes(
+	existingVolumeMounts []StepVolumeMount,
+	strategyVolumes []buildv1alpha1.BuildVolume,
+	buildVolumes []buildv1alpha1.BuildVolume,
+	buildrunVolumes []buildv1alpha1.BuildVolume,
+) ([]corev1.Volume, error) {
+	res := []corev1.Volume{}
+
+	// first we merge build volumes into the strategy ones, next we merge
+	// build run volumes into result of the previous merge.
+	// eventual list of volumes will be added to the generated TaskSpec object
+	volumes, err := MergeBuildVolumes(strategyVolumes, buildVolumes)
+	if err != nil {
+		return nil, err
+	}
+	volumes, err = MergeBuildVolumes(volumes, buildrunVolumes)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, v := range volumes {
+		volumeMountExists := false
+		for _, volumeMount := range existingVolumeMounts {
+			if volumeMount.Name == v.Name {
+				volumeMountExists = true
+
+				// In case volume mount is not read only and
+				// volume type is either secret or config map,
+				// build should not be run, because this situation may lead
+				// to errors
+				if !volumeMount.ReadOnly &&
+					(v.VolumeSource.ConfigMap != nil ||
+						v.VolumeSource.Secret != nil) {
+					return nil,
+						fmt.Errorf("Volume Mount %q must be read only", v.Name)
+				}
+			}
+		}
+
+		if !volumeMountExists && v.Optional != nil && !*v.Optional {
+			return nil, fmt.Errorf("Volume %q is not optional and does not have corresponding Volume Mount", v.Name)
+		}
```

Comment:
I assume, `optional` was meant in a different way. Basically the strategy author defines everything, the volumes and the volumeMounts on the steps. It does not make sense to define a strategy with a volume without using it in a volumeMount. There is nothing that could add a step with a volumeMount. Therefore, optional does not make sense that way.

I think with `optional`, we had something else in mind: the build strategy author (especially the cluster build strategy author) defines a volumeMount, for example for BuildKit's cache directory. The build user can mount in a persistent volume, or not. And now I am not clear on how we want to use `optional`.

* Option 1: We could ask the build strategy author to define an `emptyDir` as volume and mark it overridable. No need for an `optional` flag on the volume. The TaskRun/Pod will always contain the volume.
* Option 2: We could ask the build strategy author to define a volume with `optional` set to true. We then need to make the volumeSource optional. If the Build or BuildRun defines the volume, then this will be used. If it is present, volume and volumeMount will be in the TaskRun/Pod. If the Build or BuildRun don't define the volume, then we must decide what to do - either fall back to an emptyDir or completely omit the volume and volumeMount(s).

@adambkaplan fya

Reply:
Yeah, I had a few questions about the `optional`, but forgot to raise the question. Is it possible to have a quick call tomorrow? I would prefer going with `Option 1`, but we can also go with `Option 2`, but it needs to be documented well, I think. It feels a bit counterintuitive to me.